hopsworks-common/src/main/java/io/hops/hopsworks/common/featurestore/query/ConstructorController.java

```diff
@@ -0,0 +1,387 @@
+import org.apache.calcite.sql.SqlNodeList;
+import org.apache.calcite.sql.SqlSelect;
+import org.apache.calcite.sql.dialect.SparkSqlDialect;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.sql.parser.SqlParserPos;
+
+import javax.ejb.EJB;
+import javax.ejb.Stateless;
+import javax.ejb.TransactionAttribute;
+import javax.ejb.TransactionAttributeType;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Stateless
+@TransactionAttribute(TransactionAttributeType.NEVER)
+public class ConstructorController {
+
+  @EJB
+  private FeaturegroupFacade featuregroupFacade;
+  @EJB
+  private FeaturestoreFacade featurestoreFacade;
+  @EJB
+  private FeaturegroupController featuregroupController;
+
+  private final static String ALL_FEATURES = "*";
+
+  public ConstructorController() { }
+
+  // For testing
+  protected ConstructorController(FeaturegroupController featuregroupController,
+                                  FeaturestoreFacade featurestoreFacade,
+                                  FeaturegroupFacade featuregroupFacade) {
+    this.featuregroupController = featuregroupController;
+    this.featurestoreFacade = featurestoreFacade;
+    this.featuregroupFacade = featuregroupFacade;
+  }
+
+  public String construct(QueryDTO queryDTO) throws FeaturestoreException {
+    Query query = convertQueryDTO(queryDTO, 0);
+    // Generate SQL
+    return generateSQL(query);
+  }
+
+  /**
+   * Recursively convert the QueryDTO into the internal query representation
+   * @param queryDTO
+   * @param fgId each feature group will be aliased as fg[Integer] where [integer] is going to be an incremental id
+   * @return
+   */
+  protected Query convertQueryDTO(QueryDTO queryDTO, int fgId) {
+    Featuregroup fg = validateFeaturegroupDTO(queryDTO.getLeftFeatureGroup());
+    String fgAs = generateAs(fgId++);
+
+    String featureStore = featurestoreFacade.getHiveDbName(fg.getFeaturestore().getHiveDbId());
+    List<FeatureDTO> availableFeatures = featuregroupController.getFeatures(fg);
+    List<FeatureDTO> requestedFeatures = validateFeatures(fg, fgAs, queryDTO.getLeftFeatures(), availableFeatures);
+
+    Query query = new Query(featureStore, fg, fgAs, requestedFeatures, availableFeatures);
+    // If there are any join, recursively conver the Join's QueryDTO into the internal Query representation
+    if (queryDTO.getJoins() != null && !queryDTO.getJoins().isEmpty()) {
+      query.setJoins(convertJoins(query, queryDTO.getJoins(), fgId));
+    }
+
+    return query;
+  }
+
+  private String generateAs(int id) {
+    return "fg" + id;
+  }
+
+  /**
+   * Validate FeatureGroupDTO to make sure it exists. Authorization is done at the storage layer by HopsFS when
+   * actually executing the query.
+   * @param featuregroupDTO
+   * @return
+   */
+  private Featuregroup validateFeaturegroupDTO(FeaturegroupDTO featuregroupDTO) {
```

Comment:
The name is misleading as we are just getting the featuregroup by id from the db.
We have a function for that i[n the featuregroup controller already](https://github.com/logicalclocks/hopsworks/blob/e4f993d1e880fe44f461e6c47e32c755e478ec50/hopsworks-common/src/main/java/io/hops/hopsworks/common/featurestore/featuregroup/FeaturegroupController.java#L615) can we use this one, to avoid duplicate code?

Reply:
The other method requires passing (so fetching it before) the Featurestore object.